packages/vue-urql/src/useQuery.ts

```diff
@@ -69,80 +82,106 @@ export function useQuery<T = any, V = object>(
     createRequest<T, V>(args.query, args.variables as V) as any
   );
 
-  watchEffect(
+  const source: Ref<Source<OperationResult<T, V>> | undefined> = ref();
+  const unsubscribe: Ref<() => void> = ref(noop);
+
+  watch(
+    [args.query, args.variables],
     () => {
       const newRequest = createRequest<T, V>(args.query, args.variables as any);
       if (request.value.key !== newRequest.key) {
         request.value = newRequest;
       }
     },
     {
+      immediate: true,
       flush: 'sync',
     }
   );
 
-  let isThenable = false;
-
-  const executeQuery = (opts?: Partial<OperationContext>) => {
-    fetching.value = true;
-
-    const query$ = pipe(
-      client.executeQuery<T, V>(request.value, {
-        requestPolicy: args.requestPolicy,
-        pollInterval: args.pollInterval,
-        ...args.context,
-        ...opts,
-      }),
-      onEnd(() => {
-        fetching.value = false;
-      }),
-      onPush(res => {
-        isThenable = false;
-        data.value = res.data;
-        stale.value = !!res.stale;
-        fetching.value = false;
-        error.value = res.error;
-        operation.value = res.operation;
-        extensions.value = res.extensions;
-      }),
-      share
-    );
-
-    if (!isThenable) {
-      unsubscribe.value();
-      unsubscribe.value = publish(query$).unsubscribe;
+  watch(
+    [isPaused, request, args.requestPolicy, args.pollInterval, args.context],
+    () => {
+      if (!isPaused.value) {
+        source.value = pipe(
+          client.executeQuery<T, V>(request.value, {
+            requestPolicy: args.requestPolicy,
+            pollInterval: args.pollInterval,
+            ...args.context,
+          }),
+          share
+        );
+      } else {
+        source.value = undefined;
+      }
+    },
+    {
+      immediate: true,
+      flush: 'sync',
     }
+  );
 
-    return {
-      then(onFulfilled, onRejected) {
-        return pipe(query$, take(1), toPromise).then(onFulfilled, onRejected);
-      },
-    } as PromiseLike<OperationResult<T, V>>;
-  };
-
-  watchEffect(
-    onInvalidate => {
-      if (!isPaused.value) {
-        executeQuery();
-        onInvalidate(() => {
-          unsubscribe.value();
-        });
+  watch(
+    [source],
+    (_, __, onInvalidate) => {
+      if (!source.value) {
+        return unsubscribe.value();
```

Comment:
Maybe I'm missing something, but the `onInvalidate` callback should already take care of unsubscribing in either case. we can just return here without unsubscribing again.

Reply:
this is here because of pausing. `executeQuery` can update `source.value` and the watcher may set it to `undefined` when everything's paused, so that `executeQuery` can explicitly take precedence over the watcher's subscription being paused.